### PR TITLE
Signup

### PR DESCRIPTION
**Pull request by tunv**
Task-32886: SignUp Feature

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+class UsersController < ApplicationController
+
+  def show
+    @user = User.find(params[:id])
```
Comment (VoTaiTri) [inline app/controllers/users_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  def new
+  	@user = User.new
```
Comment (VoTaiTri) [inline app/controllers/users_controller.rb]: indent

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  end
+
+  def create
+    @user = User.new(user_params)    # Not the final implementation!
+    if @user.save
+      redirect_to @user
+      # Handle a successful save.
```
Comment (VoTaiTri) [inline app/controllers/users_controller.rb]: remove comment

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    else
+      render 'new'
+    end
+  end
+
+  private
+	def user_params
```
Comment (VoTaiTri) [inline app/controllers/users_controller.rb]:
```ruby
private
def user_params
  params.require(:user).permit :name, :email, :password, :password_confirmation
end

File: app/views/users/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,21 @@
+<% provide(:title, 'Sign up') %>
+<h1>Sign up</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user) do |f| %>
```
Comment (VoTaiTri) [inline app/views/users/new.html.erb]:
```ruby
form_for @user

File: app/views/users/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,21 @@
+<% provide(:title, 'Sign up') %>
```
Comment (VoTaiTri) [inline app/views/users/new.html.erb]: I18n
same below

File: app/views/users/show.html.erb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<% provide(:title, @user.name) %>
```
Comment (VoTaiTri) [inline app/views/users/show.html.erb]: remove ()

File: config/routes.rb [hunk 8]
```diff
@@ -1,5 +1,10 @@
 Rails.application.routes.draw do
-  root "static_pages#home"
 
+  root "static_pages#home"
+  
   get "help" => "static_pages#help"
+  get 'users/new'
```
Comment (VoTaiTri) [inline config/routes.rb]: remove this line

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+  	@user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)    # Not the final implementation!
+    if @user.save
+      redirect_to @user
+      # Handle a successful save.
+    else
+      render 'new'
```
Comment (VoTaiTri) [inline app/controllers/users_controller.rb]: user " instead '

File: config/routes.rb [hunk 10]
```diff
@@ -1,5 +1,10 @@
...
+  get "signup"  => "users#new"
+
+  resources :users
```
Comment (VoTaiTri) [inline config/routes.rb]:
```ruby
resources :users, only: [:show, :new, :create]

File: app/views/users/new.html.erb [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+      <%= f.label :name %>
+      <%= f.text_field :name %>
+      <%= f.label :email %>
+      <%= f.email_field :email %>
+      <%= f.label :password %>
+      <%= f.password_field :password %>
+      <%= f.label :password_confirmation, "Confirmation" %>
+      <%= f.password_field :password_confirmation %>
+
+      <%= f.submit "Create my account", class: "btn btn-success" %>
```
Comment (VoTaiTri) [inline app/views/users/new.html.erb]: I18n

File: app/views/users/new.html.erb [hunk 12]
```diff
@@ -0,0 +1,21 @@
+<% provide(:title, 'Sign up') %>
+<h1>Sign up</h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user) do |f| %>
+      <%= render 'shared/error_messages' %>
```
Comment (VoTaiTri) [inline app/views/users/new.html.erb]: "

File: app/views/users/new.html.erb [hunk 13]
```diff
@@ -0,0 +1,21 @@
+<% provide(:title, 'Sign up') %>
+<h1>Sign up</h1>
```
Comment (VoTaiTri) [inline app/views/users/new.html.erb]: I18n

File: app/views/shared/_error_messages.html.erb [hunk 14]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize(@user.errors.count, "error") %>.
```
Comment (VoTaiTri) [inline app/views/shared/_error_messages.html.erb]: I18n

File: app/helpers/users_helper.rb [hunk 15]
```diff
@@ -0,0 +1,8 @@
+module UsersHelper
+	
+  def gravatar_for(user)
```
Comment (VoTaiTri) [inline app/helpers/users_helper.rb]:
```ruby
gravatar_for user
```

File: app/controllers/users_controller.rb [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+  def show
+    @user = User.find(params[:id])
+  end
+
+  def new
+  	@user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)    # Not the final implementation!
```
Comment (VoTaiTri) [inline app/controllers/users_controller.rb]: remove () and comment

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+  def show
+    @user = User.find(params[:id])
+  end
+
+  def new
+  	@user = User.new
+  end
+
+  def create
+    @user = User.new(user_params)
```
Comment (VoTaiTri) [inline app/controllers/users_controller.rb]: ()

File: app/views/users/new.html.erb (hunk 13 above)
Comment (tunv) [inline app/views/users/new.html.erb]: Cái này em không dùng I18n được ạ!

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new user_params
+    if @user.save
+      redirect_to @user
+    else
+      render 'new'
```
Comment (VoTaiTri) [inline app/controllers/users_controller.rb]:
```ruby
render "new"

File: app/views/shared/_error_messages.html.erb [hunk 19]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      The form contains <%= pluralize(@user.errors.count, I18n.t("error")) %>.
```
Comment (VoTaiTri) [inline app/views/shared/_error_messages.html.erb]: I18n for text : "The form contains"

File: app/controllers/users_controller.rb [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+    if @user.save
+      redirect_to @user
+    else
+      render "new"
+    end
+  end
+
+  private
+  def user_params
+  params.require(:user).permit :name, :email, :password, :password_confirmation
```
Comment (VoTaiTri) [inline app/controllers/users_controller.rb]:
```ruby
def user_params
  params.require(:user).permit :name, :email, :password, :password_confirmation
end

Comment (tunv) [inline app/controllers/users_controller.rb]: Edited

File: app/views/shared/_error_messages.html.erb [hunk 21]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+  <div id="error_explanation">
+    <div class="alert alert-danger">
+      <%= t("page_error")%> <%= pluralize(@user.errors.count, I18n.t("error")) %>.
```
Comment (VoTaiTri) [inline app/views/shared/_error_messages.html.erb]: space before %>
```ruby
<%= t("page_error") %>

File: app/views/users/new.html.erb [hunk 22]
```diff
@@ -0,0 +1,21 @@
+<% provide :title, t("signup") %>
+<h1><%= t "signup"%></h1>
```
Comment (VoTaiTri) [inline app/views/users/new.html.erb]:
```ruby
<%= t("signup") %>

File: app/views/users/new.html.erb [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+    <%= form_for @user do |f| %>
+      <%= render "shared/error_messages" %>
+    
+      <%= f.label :name %>
+      <%= f.text_field :name %>
+      <%= f.label :email %>
+      <%= f.email_field :email %>
+      <%= f.label :password %>
+      <%= f.password_field :password %>
+      <%= f.label :password_confirmation, I18n.t("confirm") %>
```
Comment (VoTaiTri) [inline app/views/users/new.html.erb]: nt

File: app/views/users/new.html.erb [hunk 24]
```diff
@@ -0,0 +1,21 @@
...
+      <%= f.password_field :password_confirmation %>
+
+      <%= f.submit I18n.t("create_acount"), class: "btn btn-success" %>
```
Comment (VoTaiTri) [inline app/views/users/new.html.erb]:
```ruby
<%= f.submit t("create_acount"), class: "btn btn-success" %>

Comment (tunv) [inline app/views/users/new.html.erb]: Edited

**Comment by VoTaiTri** [comment]
OK

File: app/views/users/new.html.erb [hunk 25]
```diff
@@ -0,0 +1,21 @@
+<% provide :title, t("signup") %>
+<h1><%= t "signup" %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for @user do |f| %>
```
Comment (dieunb) [inline app/views/users/new.html.erb]: Slice form to new partial file

File: app/views/users/new.html.erb [hunk 26]
```diff
@@ -0,0 +1,8 @@
+<% provide :title, t("signup") %>
+<h1><%= t "signup" %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+      <%= render "users/form" %>
```
Comment (VoTaiTri) [inline app/views/users/new.html.erb]:
```ruby
<%= render "form" %>

**Pull request merged**